drivers/siemens-sitop.c

```diff
@@ -0,0 +1,314 @@
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "BA<85", 5) == 0) {
+					current_ups_status.battery_above_85_percent = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else {
+					/* nothing sensible found at the start of the rx_buffer. */
+					rm_buffer_head(1);
+				}
+			}
+		}
+	}
+	return new_data_received;
+}
+
+
+static int instcmd(const char *cmdname, const char *extra) {
+	/* See the remarks in upsdrv_shutdown() */
+	if (!strcasecmp(cmdname, "shutdown.return")) {
+		upslogx(LOG_NOTICE, "instcmd: sending command R");
+		ser_send_pace(upsfd, 200000, "\n\nR\n\n");
+		ser_send_pace(upsfd, 200000, "R\n\n");
+		return STAT_INSTCMD_HANDLED;
+	}
+	if (!strcasecmp(cmdname, "shutdown.stayoff")) {
+		upslogx(LOG_NOTICE, "instcmd: sending command S");
+		ser_send_pace(upsfd, 200000, "\n\nS\n\n");
+		ser_send_pace(upsfd, 200000, "S\n\n");
+		return STAT_INSTCMD_HANDLED;
+	}
+
+	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
+	return STAT_INSTCMD_UNKNOWN;
+}
+
+
+void upsdrv_initinfo(void) {
+	int max_attempts = 5;
+	int found = 0;
+	while (!found && max_attempts > 0) {
+		if (check_for_new_data())
+			found = 1;
+		else
+			sleep(1); /* Sleep a while, then try again */
+		max_attempts--;
+	}
+	if (!found)
+		fatalx(EXIT_FAILURE, "No data received from the UPS");
+	
+	dstate_setinfo("ups.mfr", "Siemens");
+	dstate_setinfo("ups.model", "SITOP UPS500 series");
+	/* note: for a transition period, these data are redundant! */
+	dstate_setinfo("device.mfr", "Siemens");
+	dstate_setinfo("device.model", "SITOP UPS500 series");
+	
+	/* supported commands: */
+	dstate_addcmd("shutdown.stayoff");
+	dstate_addcmd("shutdown.return");
+
+	upsh.instcmd = instcmd;
+}
+
+
+void upsdrv_updateinfo(void) {
+	if (check_for_new_data())
+		nr_polls_without_data = 0;
+	else {
+		nr_polls_without_data++;
+		if (nr_polls_without_data < 0)
+			nr_polls_without_data = INT_MAX;
+	}
+	
+	if (nr_polls_without_data > max_polls_without_data) {
+		/* data is stale */
+		dstate_datastale();
+		return;
+	}
+	
+	/* This is all we know about the charge level... */
+	dstate_setinfo("battery.charge.indication",
+		(current_ups_status.battery_above_85_percent) ? ">85%%" : "<85%%");
+	
+	status_init();
+	
+	if (current_ups_status.dc_input_low || current_ups_status.on_battery) {
+		status_set("OB");
+	} else {
+		status_set("OL");
+	}
+	if (current_ups_status.battery_alarm) {
+		status_set("LB");
+	}
+	
+	status_commit();
+	dstate_dataok();
+	ser_comm_good();
+
+	/*
+	 * Fast polling is preferred, because
+	 * A) the UPS spits out new data every 75 msec,
+	 * B) some models in this SITOP series have a _very_ small capacity
+	 *    (< 10sec runtime), so every second might count.
+	 */
+	poll_interval = 1;
+}
+
+void upsdrv_shutdown(void) {
+	/* tell the UPS to shut down, then return - DO NOT SLEEP HERE */
+	upslogx(LOG_NOTICE, "instcmd: sending command R");
+
+	/* Note: the UPS does not really like to receive data.
```

Comment:
Why not moving these remarks in `instcmd()` and simply call `instcmd("shutdown.return", NULL)` here?

Reply:
Yes, better.